### ISSUE TITLE
Learnings from CerebrasResearch/reap: ranking-signal collection, test methodology, and a negative prefill-ranking POC

### ISSUE THREAD

**Issue by ddvnguyen** [ddvnguyen/hydra_vortex#772]
## Summary

Studied [CerebrasResearch/reap](https://github.com/CerebrasResearch/reap) (Router-weighted Expert Activation Pruning) at owner direction. REAP and the `baseline-flash-next` expert-residency track share a core: **track expert usage, rank experts**. Three transferable learnings, one of which supersedes an existing methodology ruling, plus a POC that was run immediately and returns a **negative result** on prefill-derived ranking.

Repo studied at `/tmp/reap-study` (shallow clone).

---

## Learning 1 — collect all ranking signals in ONE pass, choose the ranking offline

`src/reap/pruning_metrics.py:22-64` accumulates **seven** per-layer signals side by side:

| REAP state key | Meaning | Hydra has it? |
|---|---|---|
| `expert_frequency` | raw selection count | ✅ **our only metric** |
| `weighted_expert_frequency_sum` | Σ router gate weight ("soft count") | ❌ |
| `ean_sum` / `ean_mean` | Σ / mean ‖activation‖ | ❌ |
| `weighted_ean_sum` | Σ (gate × ‖act‖) | ❌ |
| `reap` | **mean** (gate × ‖act‖) — the REAP score | ❌ |
| `max_activations` | peak ‖act‖ | ❌ |
| `pairwise_expert_frequency` | fᵢ+fⱼ marginal-sum matrix | ❌ |

REAP score is one line (`pruning_metrics.py:198`):
```python
reap[i] = (ean_norm * active_router_weights).mean()
```

**The lesson is structural, not the formula.** They collect everything once and ablate rankings offline for free. Our `.route` traces record selection only — every weight token is literally `:1` — so trying any other ranking requires a new rig campaign.

Caveat for whoever picks this up: `pairwise_expert_frequency` is `fᵢ+fⱼ`, a marginal sum used as a merge denominator. It is **not** a co-activation matrix. Do not mistake it for one.

**Action:** widen the route-trace emitter to record router gate value (and ideally ‖activation‖) alongside the expert id.

---

## Learning 2 — test methodology (highest-value reusable part)

**(a) Closed-form unit tests on hand-computable tensors** — `tests/test_pruning_metrics.py`. Activations are chosen as Pythagorean triples (`[3,4]→5`, `[5,12]→13`, `[8,15]→17`, `[7,24]→25`) so every norm is an exact integer and expected saliency is verifiable by hand. Two tests, 157 lines, fully cover masking + renormalization.

Our h-estimators and pin scorers currently have **zero unit tests**. This program has already absorbed two instrument errors (the T-2 `cpu_fraction` misread; the G3 liveness void where `hit=miss=0` on all 15 legs). This pattern is cheap and catches exactly that class.

**(b) Tiny synthetic models for plumbing e2e** — `tests/test_pruning_e2e.py:62`: `hidden_size=16, num_hidden_layers=3, num_experts=3, vocab_size=32`, parametrized across four architectures, asserting structural invariants. CPU-instant, no GPU, no real model.

**(c) Teacher-forced per-position divergence** — `scripts/generation_quality_analysis.py:94`. Models are compared by **one forward pass over a FIXED token sequence**, computing per-position forward-KL / reverse-KL / JSD / Wasserstein between base and candidate logits. Not by comparing generated text.

---

## Learning 3 — (c) supersedes the Phase-B correctness gate, and we already ship the instrument

The Stage-1 bit-identity gate went VOID because same-binary controls diverge at char 30 — the rig is nondeterministic across processes, so cross-process sha-equality can never pass. The fallback ruling was "divergence-point as screen, PPL as binding gate."

**PPL is the wrong statistic.** Teacher-forced per-position KL is strictly better:

1. It sidesteps the nondeterminism. Char-30 divergence comes from *sampled trajectories* diverging; holding the token sequence fixed leaves only FP reduction-order noise.
2. It yields a **distribution**, not a scalar — real power at small n. PPL is a scalar summary with poor power.
3. It still localizes *where* divergence starts (what divergence-point was wanted for) but continuously and against a null.

**It is already in our fork:**
```
common/arg.cpp:2502   --kl-divergence
common/arg.cpp:2509   --save-all-logits / --kl-divergence-base FNAME
```
`tools/perplexity/perplexity.cpp:1949-2002` emits Mean KLD ± σ, the full KLD percentile ladder (0.1/1/5/10/50/90/95/99/99.9/max), Δp RMS, and Same-top-p.

**Revised Phase-B bar:** dump baseline logits once with `--save-all-logits`; run A′ against it to get the FP-noise null ladder; run each candidate against the same base; PASS = candidate KLD ladder within the A/A′ null, Same-top-p at null level. Binding statistic = median + 99th-percentile KLD.

**Unverified precondition:** `--kl-divergence` has not been confirmed working against a `qwen4exp` GGUF on the rig. ~10 min check required before acting.

---

## Learning 4 — calibration-mix DSL

`src/reap/data.py:117` — composable, versionable dataset spec:
```
theblackcat102/evol-codealpaca-v1:4096
open-r1/Mixture-of-Thoughts[code]:4096
SWE-bench/SWE-smith-trajectories(tool):4096
dataset[subset](split):num_batches
```
This is the machine-readable equivalent of our hand-maintained 30-row `V4_SET_TABLE.md`.

---

## Correction to a previously recorded risk

Earlier assessment said "arch port effort unknown." For the **observer** side it is near-zero — one regex in a dataclass (`observer.py:483`):
```python
class Qwen3MoEObserverHookConfig(MoETransformerObserverConfig):
    module_class_name_to_hook_regex = "Qwen3MoeSparseMoeBlock"
```
Requires only that the module expose `num_experts`/`top_k` and return `router_logits`. The **pruning** side needs per-arch model surgery and is a real project.

---

## POC run: prefill-derived ranking — NEGATIVE RESULT

Hypothesis tested: rank experts during **prefill**, so a high-ranked pin set exists before decode token 0. Motivation was that Phase-1b killed prefill ranking at `P=79` / `P=8`, with a length curve (breakeven k≈48-64) still climbing at k=64 — while production prefills are far longer.

Ran on **all 57 traces** (v2 + v4, prefill 79→7546 tokens), CPU replay, zero GPU. N=38, `S(h)=1/(1−0.545h+0.1298)`.

| arm | median h | mean | min | max | S(median) |
|---|---|---|---|---|---|
| `h_prefill` (own prefill → own decode) | **0.2383** | 0.2496 | 0.1268 | 0.4250 | **1.000** |
| `h_global` (LOSO pooled decode) | 0.3363 | 0.3172 | 0.1349 | 0.4118 | 1.057 |
| `h_subject` (LOSO within subject) | 0.3787 | 0.3672 | 0.1562 | 0.4661 | 1.083 |
| `h_oracle` (in-sample decode top-38) | 0.4984 | 0.5051 | 0.4327 | 0.6554 | 1.165 |

Paired `prefill − global`: median **−0.0926**, positive in only **13/57** traces.

**Verdict: prefill ranking stays dead.** It lands exactly on breakeven (h*=0.2382) and loses to the deployable global table.

**Why the length-curve extrapolation failed (a real methodological trap):** the Phase-1b curve varied `k` as a *truncation of one fixed prompt* — more tokens = better estimate of that prompt. Production varies prompt length *across different prompts*, and in v4 longer prefill also means a later session turn carrying prior turns. Measured direction is the **opposite** of the curve's: the two shortest prompts score best (`c01` P=79 → 0.4064; `c02` P=122 → 0.4250) and the longest score worst (`coding_S2_T2` P=5010 → 0.1268; `coding_S4_T2` P=7299 → 0.1607). **A within-prompt length curve does not extrapolate to across-prompt length.**

### Two findings from the POC that outrank the prefill question

1. **Offline `h_global` = 0.336 but the rig measured h = 0.14 on prompt767.** A 2.4× gap between the offline estimator and the live instrument. One of them is wrong, and this now gates the entire ranking program — if live should be 0.336 then `S = 1.057` and the path is alive; if offline is inflated then every h number in the program is suspect. **Highest-priority diagnostic.**
2. **The whole achievable band at N=38 is thin.** Even the *oracle* reaches only `S = 1.165`. SHIP is 1.10. There is very little room between "perfect ranking" and "not worth building" at this operating point — which argues the lever is N or the cost model, not the ranking function.

Artifacts: `/tmp/opencode/prefill_rank_poc.py`, `/tmp/opencode/prefill_score.py`, `/tmp/opencode/prefill_poc_results.json`.

---

## Proposed actions (priority order)

1. **Diagnose the 0.336 vs 0.14 estimator/instrument gap.** Blocks interpretation of everything else.
2. **Repoint Phase-B onto `--kl-divergence`** against an A/A′ null ladder (after the 10-min qwen4exp feasibility check).
3. **Widen the route-trace emitter** to record gate value + ‖activation‖, not just selection.
4. **Unit-test the h-scorers** with closed-form tiny-tensor fixtures.
5. **Adopt the dataset-spec DSL** for the subject corpus.
6. **Park REAP pruning itself** — it is a model change, Stage-1 is an engine change; orthogonal, and it needs a quality gate, which (2) provides.

## Explicitly NOT adopting

REAP's vLLM + lm-eval + EvalPlus + LiveCodeBench stack (`src/reap/eval.py:12-17`). Wrong runtime for a llama.cpp/GGUF shop on 12 GB. Take the methodology, not the dependencies.
